### PR TITLE
[mlir][py] Overload print with state.

### PR DESCRIPTION
**Pull request by jpienaar**
Enables reusing the AsmState when printing from Python. Also moves the fileObject and binary to the end (pybind11::object was resulting in the overload not working unless `state=` was specified).

### CODE REVIEW

File: mlir/lib/Bindings/Python/IRCore.cpp [hunk 1]
```diff
@@ -1198,6 +1207,17 @@ void PyOperationBase::print(py::object fileObject, bool binary,
   mlirOpPrintingFlagsDestroy(flags);
 }
 
+void PyOperationBase::print(PyAsmState &state, pybind11::object fileObject,
+                            bool binary) {
```
Comment (makslevental) [inline mlir/lib/Bindings/Python/IRCore.cpp:1211]: nittiest of nits:

```suggestion
void PyOperationBase::print(PyAsmState &state, py::object fileObject,
                            bool binary) {
```

File: mlir/lib/Bindings/Python/IRCore.cpp [hunk 2]
```diff
@@ -2946,15 +2967,23 @@ void mlir::python::populateIRCore(py::module &m) {
                                /*assumeVerified=*/false);
           },
           "Returns the assembly form of the operation.")
-      .def("print", &PyOperationBase::print,
+      .def("print",
+           py::overload_cast<PyAsmState &, pybind11::object, bool>(
```
Comment (makslevental) [inline mlir/lib/Bindings/Python/IRCore.cpp:2971]:
```suggestion
           py::overload_cast<PyAsmState &, py::object, bool>(
```

File: mlir/lib/Bindings/Python/IRCore.cpp [hunk 3]
```diff
@@ -2946,15 +2967,23 @@ void mlir::python::populateIRCore(py::module &m) {
...
+               &PyOperationBase::print),
+           py::arg("state"), py::arg("file") = py::none(),
+           py::arg("binary") = false, kOperationPrintStateDocstring)
+      .def("print",
+           py::overload_cast<std::optional<int64_t>, bool, bool, bool, bool,
+                             bool, pybind11::object, bool>(
```
Comment (makslevental) [inline mlir/lib/Bindings/Python/IRCore.cpp:2977]:
```suggestion
                             bool, py::object, bool>(
```

File: mlir/lib/Bindings/Python/IRModule.h [hunk 4]
```diff
@@ -550,16 +550,19 @@ class PyModule : public BaseContextObject {
...
 public:
   virtual ~PyOperationBase() = default;
   /// Implements the bound 'print' method and helps with others.
-  void print(pybind11::object fileObject, bool binary,
-             std::optional<int64_t> largeElementsLimit, bool enableDebugInfo,
+  void print(std::optional<int64_t> largeElementsLimit, bool enableDebugInfo,
              bool prettyDebugInfo, bool printGenericOpForm, bool useLocalScope,
-             bool assumeVerified);
+             bool assumeVerified, pybind11::object fileObject, bool binary);
+  void print(PyAsmState &state, pybind11::object fileObject, bool binary);
```
Comment (makslevental) [inline mlir/lib/Bindings/Python/IRModule.h:564]:
```suggestion
             bool assumeVerified, py::object fileObject, bool binary);
  void print(PyAsmState &state, py::object fileObject, bool binary);
```

**Review by makslevental — APPROVED** [review]
LGTM. I'm not sure that the test actually tests the intent (_reusing_ `AsmState`) but I don't think it matters (it sufficiently exercises the code).

**Review by ftynse — APPROVED** [review]

**Comment by stellaraccident** [comment]
Thanks.

**Pull request merged**